proto_parser/src/proto_parser.cc

```diff
@@ -15,14 +15,65 @@
 */
 
 #include "proto_parser.h"
+#include "message_value.h"
+#include "primitive_value.h"
 
 namespace wireless_android_play_analytics {
-
+  
 void ProtoParser::PopulateFields(int& prev_field_line, 
     const google::protobuf::TextFormat::ParseInfoTree& tree,
     const google::protobuf::Message& message,
     ProtoValue* proto_value, int indent_count) {
-  // TODO(alexanderlin): Implement.
+  assert(proto_value != nullptr);
+  const google::protobuf::Reflection* reflection = message.GetReflection();
+  const google::protobuf::Descriptor* descriptor = message.GetDescriptor();
+  MessageValue* message_tmp = 
+      dynamic_cast<MessageValue*>(proto_value);
+  assert(message_tmp);
+  int field_count = descriptor->field_count();
+  std::vector<FieldInfo> field_info_list;
+  // Populate field_info_list with all the fields in the text proto. 
+  for (int i = 0; i < field_count; ++i) {
+    const google::protobuf::FieldDescriptor* field_descriptor = 
+        descriptor->field(i);
+    if (field_descriptor->is_repeated()) {
+      int field_size = reflection->FieldSize(message, field_descriptor);
+      for (int j = 0; j < field_size; ++j) {
+        field_info_list.push_back(FieldInfo(/*line=*/GetLocation(tree, 
+            field_descriptor, j), /*index=*/j, field_descriptor));
+      }
+    } else if(reflection->HasField(message, field_descriptor)) {
+      field_info_list.push_back(FieldInfo(/*line=*/GetLocation(tree, 
+          field_descriptor, -1), /*index=*/-1, field_descriptor));
+    }
+  }
```

Comment:
I feel this part is not needed. You can store the line number in ProtoValue and sort the constructed fields (vector of ProtoValue) by the line number property.

Reply:
what would the line number be if we were to add say self-constructed protovalue then?

Reply:
Your current logic is: TreeInfo -> FieldInfo -> sort -> Field
The logic can simplified as: TreeInfo -> Field -> sort
By integrating the content of FieldInfo (effectively the line number) to Field (ProtoValue), you can remove the intermediate step and the FieldInfo.

Reply:
The problem with that is since the fields vector is a vector of pointers I effectively would have to create a new vector of dereferenced proto values then sort it and then put it back into the fields array which makes the code not only more space inefficient but also longer. Another approach I can do is create a custom comparator that takes in two unique_ptrs and dereference and compares them, however, that will require me to pass the unique_ptr to the comparator by reference which is highly discouraged. There's no way for me to pass just the pointer since the vector itself only stores unique_ptrs.

Reply:
You may sort with a custom comparator http://www.cplusplus.com/reference/algorithm/sort/